src/v/kafka/server/group_manager.cc

```diff
@@ -1739,4 +1798,96 @@ error_code group_manager::validate_group_status(
     return error_code::not_coordinator;
 }
 
+ss::future<cluster::get_producers_reply>
+group_manager::get_group_producers_locally(
+  cluster::get_producers_request request) {
+    const auto& ntp = request.ntp;
+    cluster::get_producers_reply reply;
+    auto it = _partitions.find(ntp);
+    if (it == _partitions.end() || !it->second->partition->is_leader()) {
+        reply.error_code = cluster::tx::errc::not_coordinator;
+        co_return reply;
+    }
+    auto attached_partition = *it;
+    reply.error_code = cluster::tx::errc::none;
+    // snapshot the list of groups attached to this partition
+    chunked_hash_map<group_id, group_ptr> groups;
+    std::copy_if(
+      _groups.begin(),
+      _groups.end(),
+      std::inserter(groups, groups.end()),
+      [&ntp](auto g_pair) {
+          const auto& [group_id, group] = g_pair;
+          return group->partition()->ntp() == ntp;
+      });
+    reply.producer_count = std::accumulate(
+      _groups.begin(),
+      _groups.end(),
+      size_t(0),
+      [](size_t acc, const auto& entry) {
+          return acc + entry.second->producers().size();
+      });
+    for (auto& [gid, group] : groups) {
+        if (reply.producers.size() > request.max_producers_to_include) {
+            break;
+        }
+        if (group->in_state(group_state::dead)) {
+            continue;
+        }
+        auto partition = group->partition();
+        if (!partition) {
+            // unlikely, conservative check
+            continue;
+        }
+        for (const auto& [id, state] : group->producers()) {
+            cluster::producer_state_info producer_info;
+            producer_info.pid = {id, state.epoch};
+            producer_info.group_id = group->id()();
+            auto& tx = state.transaction;
+            if (tx) {
+                producer_info.tx_begin_offset = tx->begin_offset;
+                producer_info.tx_seq = tx->tx_seq;
+                producer_info.tx_timeout = tx->timeout;
+                auto time_since_last_update = model::timeout_clock::now()
+                                              - tx->last_update;
+                auto last_update_ts = model::timestamp_clock::now()
+                                      - time_since_last_update;
+                producer_info.last_update = model::timestamp{
+                  last_update_ts.time_since_epoch() / 1ms};
+                producer_info.coordinator_partition = tx->coordinator_partition;
+            }
+            reply.producers.push_back(std::move(producer_info));
+            if (reply.producers.size() > request.max_producers_to_include) {
```

Comment:
analogously, strictly speaking, shouldn't the condition be `>=` and be checked at the beginning of the loop body?